lib/helper.js

```diff
@@ -0,0 +1,218 @@
+                    `${stepName} step in cannot be found in current template or ${oldJob.template} template; skipping`
+                );
+            }
+        }
+
+        newJob.steps = mergedSteps;
+    // Basic step merge with template
+    } else if (fromTemplate && oldJob.steps) {
+        let stepName;
+        let preStepName;
+        let postStepName;
+        const mergedSteps = [];
+        const teardownSteps = [];
+
+        // convert steps from oldJob from array to object for faster lookup
+        const oldSteps = oldJob.steps.reduce((obj, item) => {
+            const key = Object.keys(item)[0];
+
+            if (key.startsWith('teardown-')) {
+                teardownSteps.push(key);
+            }
+
+            obj[key] = item[key];
+
+            return obj;
+        }, {});
+
+        for (let i = 0; i < newJob.steps.length; i += 1) {
+            [stepName] = Object.keys(newJob.steps[i]);
+            preStepName = `pre${stepName}`;
+            postStepName = `post${stepName}`;
+
+            // add pre-step
+            if (oldSteps[preStepName]) {
+                mergedSteps.push({ [preStepName]: oldSteps[preStepName] });
+            }
+
+            // if user doesn't define the same step, add it
+            if (!oldSteps[stepName]) {
+                mergedSteps.push(newJob.steps[i]);
+            } else if (!stepName.startsWith('teardown-')) {
+                // if user defines the same step, only add if it's not teardown
+                // otherwise, skip (it will be overriden later, otherwise will get duplicate steps)
+                mergedSteps.push({ [stepName]: oldSteps[stepName] });
+            }
+
+            // add post-step
+            if (oldSteps[postStepName]) {
+                mergedSteps.push({ [postStepName]: oldSteps[postStepName] });
+            }
+        }
+
+        for (let i = 0; i < teardownSteps.length; i += 1) {
+            stepName = teardownSteps[i];
+            mergedSteps.push({ [stepName]: oldSteps[stepName] });
+        }
+
+        newJob.steps = mergedSteps;
+    }
+
+    return warnings;
+}
+
+/**
+ * Retrieve template and merge into job config
+ * @method mergeTemplateIntoJob
+ * @param  {Object}           templateObj       Template object with job config
+ * @param  {TemplateFactory}  templateFactory   Template Factory to get template from
+ * @return {Promise}                            Resolves with obj with:
+ *                                              - new flattened job config
+ *                                              - parent template images object
+ */
+async function mergeTemplateIntoJob(templateObj, templateFactory) {
+    // Child template config
+    const oldJob = templateObj.config;
+
+    // Try to get the template
+    return templateFactory.getTemplate(oldJob.template)
+        .then((template) => {
+            if (!template) {
+                throw new Error(`Template ${oldJob.template} does not exist`);
+            }
```

Comment:
here we can also check child template's `order` against locked steps before merging two templates

Reply:
I think it'll be better to add locked steps feature in a separate PR